docs/design/1161-standardize-stdlib.md

```diff
@@ -0,0 +1,237 @@
+# Proposal: Standardizing the Alloy standard library
+
+* Author: Robert Fratto (@rfratto)
+* Last updated: 2024-06-27
+* Discussion link: <https://github.com/grafana/alloy/pull/1161>
+
+## Abstract
+
+This proposal introduces a way to standardise on how new capabilities are added
+to Alloy's standard library so that usage of the standard library feels
+consistent.
+
+## Problem
+
+As of Alloy v1.2.0, there are 19 identifiers in the standard library:
+
+* `base64_decode`
+* `coalesce`
+* `concat`
+* `constants`
+* `env`
+* `format`
+* `join`
+* `json_decode`
+* `json_path`
+* `nonsensitive`
+* `replace`
+* `split`
+* `to_lower`
+* `to_upper`
+* `trim`
+* `trim_prefix`
+* `trim_space`
+* `trim_suffix`
+* `yaml_decode`
+
+These identifiers were added organically over time; there has been little
+thought to consistency across all identifiers.
+
+While 19 identifiers is manageable and possible for a user to remember the most
+important ones, if we don't establish a way to standardise existing and new
+identifiers, the inconsistency between identifiers will only grow over time as
+new identifiers are added and become harder to understand.
+
+This proposal aims to:
+
+* Establish a convention for how to name identifiers in the standard library.
+* Propose a way to align existing identifiers with the new convention without
+  breaking backwards compatibility.
+
+For the sake of demonstrating the convention, new identifiers are provided as
+examples, but are out of scope of this proposal and should not be considered
+accepted if this proposal is accepted.
+
+## Proposal
+
+The standard library will introduce namespaces for identifiers that group
+similar identifiers together:
+
+* Array functions
+    * `array.concat` (previously `concat`)
+* Conversion functions
+    * `convert.nonsensitive` (previously `nonsensitive`)
+* Decoding functions
+    * `decode.base64` (previously `decode_base64`)
+    * `decode.json` (previously `decode_json`)
+    * `decode.yaml` (previously `decode_yaml`)
```

Comment:
I have similar concerns for the `decode` namespace as I do for os. It seems at least a little likely that we will have some encoding functionality, like the ability to encode base64 or produce a checksum of an input. 

It doesn't feel like encoding should be added into a decode-specific namespace. @mattdurham suggested a `format` namespace which could contain both encoding and decoding functionality

* `format.base64.encode`
* `format.base64.decode` 
* `format.json.decode` 
* (etc)

We could also call the namespace "encoding" (similar to Go's encoding/json package) and use different verbs to describe encoding/decoding to avoid using the word "encode" more than once:

* `encoding.to_base64` 
* `encoding.from_base64` 
* `encoding.to_json` 
* (etc)

Reply:
I would rather the lower block. The three name block feels to much, which is more stylistic since its not very different in the actual number of characters.